hatice-ecevit-2016400138/assignment2/BinaryTreeNode.java

```diff
@@ -0,0 +1,118 @@
+import java.util.ArrayList;
+public class BinaryTreeNode<T> {
+    private T data;
+    private BinaryTreeNode<T> right;
+    private BinaryTreeNode<T> left;
+
+    public BinaryTreeNode(T data, BinaryTreeNode<T> left, BinaryTreeNode<T> right) {
+        this.data = data;
+        this.left = left;
+        this.right = right;
+    }
+
+    public T getData() {
+        return data;
+    }
+
+    public BinaryTreeNode<T> getRight() {
+        return right;
+    }
+
+    public BinaryTreeNode<T> getLeft() {
+        return left;
+    }
+
+    public void setData(T data) {
+        this.data = data;
+    }
+
+    public void setRight(BinaryTreeNode<T> right) {
+        this.right = right;
+    }
+
+    public void setLeft(BinaryTreeNode<T> left) {
+        this.left = left;
+    }
+
+    /**
+     * @param key key of the node whose ancestors are going to be printed
+     * @return ancestor list of key
+     */
+    ArrayList<T> findAncestors(T key) {
+        // the point at which the key is found
+        // constructs the empty ancestor list
+        if (getData().equals(key)) {
+            return new ArrayList<T>();
+        }
+        ArrayList<T> ancestorList = null;
+        if (getLeft() != null) {
+            ancestorList = getLeft().findAncestors(key);
+        }
+        // if key is on the left of the root & the key is found
+        if (ancestorList != null) {
+            ancestorList.add(getData());
+        }
+        else {
+            if (getRight() != null) {
+                ancestorList = getRight().findAncestors(key);
+            }
+            // if key is on the right of the root & the key is found
+            if (ancestorList != null) {
+                ancestorList.add(getData());
+            }
+            else {
+                return null;
+            }
+        }
+        return ancestorList;
+    }
+
+    /**
```

Comment:
This code works. But as another exercise, think to see if you could write your solution for finding a common ancestor in terms of your code for finding all ancestors.